daal4py/sklearn/_utils.py

```diff
@@ -257,3 +266,119 @@ def get_status(self, logs=False):
+        # add n_jobs to __doc__ string if needed
+        if (
+            hasattr(self, "__doc__")
+            and isinstance(self.__doc__, str)
+            and "n_jobs : int" not in self.__doc__
+        ):
+            parameters_doc_tail = "\n    Attributes"
+            n_jobs_doc = """
+    n_jobs : int, default=None
+        The number of jobs to use in parallel for the computation.
+        ``None`` means 1 unless in a :obj:`joblib.parallel_backend` context.
+        ``-1`` means using all processors.
+        See :term:`Glossary <n_jobs>` for more details.
+"""
+            self.__doc__ = self.__doc__.replace(
+                parameters_doc_tail, n_jobs_doc + parameters_doc_tail
+            )
+        self.n_jobs = n_jobs
+
+    # add "n_jobs" parameter to signature of wrapped init
+    # if estimator doesn't originally support it
+    sig = signature(init_function)
+    original_params = list(sig.parameters.values())
+    if "n_jobs" not in list(map(lambda param: param.name, original_params)):
+        original_params.append(Parameter("n_jobs", Parameter.KEYWORD_ONLY, default=None))
+        init_with_n_jobs.__signature__ = sig.replace(parameters=original_params)
+
+    return init_with_n_jobs
+
+
+# Note: getting controller in global scope of this module is required
+# to avoid overheads by its initialization per each function call
+threadpool_controller = threadpoolctl.ThreadpoolController()
+
+
+def get_suggested_n_threads(n_cpus):
+    n_threads_map = {
+        lib_ctl.internal_api: lib_ctl.get_num_threads()
+        for lib_ctl in threadpool_controller.lib_controllers
+        if lib_ctl.internal_api != "mkl"
+    }
+    # openBLAS is limited by 128 threads by default.
+    # thus, 128 threads from openBLAS is uninformative
+    if "openblas" in n_threads_map and n_threads_map["openblas"] == 128:
+        del n_threads_map["openblas"]
+    # remove default values equal to n_cpus
+    for backend in list(n_threads_map.keys()):
+        if n_threads_map[backend] == n_cpus:
+            del n_threads_map[backend]
+    if len(n_threads_map) > 0:
+        return min(n_threads_map.values())
+    else:
+        return None
+
+
+# decorator for running of methods containing oneDAL kernels with "n_jobs"
+def run_with_n_jobs(method):
+    @wraps(method)
+    def method_wrapper(self, *args, **kwargs):
+        # threading parallel backend branch
+        if not isinstance(threading.current_thread(), threading._MainThread):
+            warn(
+                "'Threading' parallel backend is not supported by "
+                "Intel(R) Extension for Scikit-learn*. "
+                "Falling back to usage of all available threads."
+            )
+            result = method(self, *args, **kwargs)
+            return result
+        # multiprocess parallel backends branch
+        method_name = (
+            f"{self.__class__.__module__}.{self.__class__.__name__}.{method.__name__}"
+        )
```

Comment:
This will result in code that's more readable, even after auto-formatting
```suggestion
        cls = self.__class__
        method_name = ".".join([cls.__module__, cls.___name__, method.__name__])
```

Reply:
Done. Btw, VSCode thinks `cls` is reserved word, so changed it to `cl`